docs/authentication.md

```diff
@@ -0,0 +1,318 @@
+    return { props: {} };
+  }
+
+  // If there is a user, return the current session
+  return { props: { session } };
+}
+```
+
+An advantage of this pattern is preventing a flash of unauthenticated content before redirecting. It's important to note fetching user data in `getServerSideProps` will block rendering until the request to your authentication provider resolves. To prevent creating a bottleneck and decreasing your TTFB ([Time to First Byte](https://web.dev/time-to-first-byte/)), you should ensure your authentication lookup is fast. Otherwise, consider static generation.
+
+There is an [open RFC](https://github.com/vercel/next.js/discussions/14890) to improve redirecting inside `getServerSideProps`.
+
+## Providers
+
+Now that we've discussed authentication patterns, let's look at specific providers and explore how they're used with Next.js.
+
+### Firebase
+
+When using [Firebase Authentication](/blog/nextjs-firebase-serverless), we recommend using the static generation pattern. There currently isn't a way to log-in server-side using the Firebase Admin SDK.
+
+It is possible to use the Firebase Client SDK to generate an ID token and forward it directly to Firebase's REST API on the server to log-in. However, requests to Firebase might take some time to resolve, depending on your user's location.
+
+Let's look an example creating a `useUser` hook. This hook allows you to fully control where you'll call `signinWithGitHub`, or any other authentication provider with Firebase.
+
+```jsx
+import React, { useState, useEffect, useContext, createContext } from 'react';
+import firebase from './firebase';
+
+const authContext = createContext();
+
+// You can wrap your _app.js with this provider
+export function AuthProvider({ children }) {
+  const auth = useProvideAuth();
+  return <authContext.Provider value={auth}>{children}</authContext.Provider>;
+}
+
+// Custom React hook to access the context
+export const useAuth = () => {
+  return useContext(authContext);
+};
+
+function useProvideAuth() {
+  // Store the user in state
+  const [user, setUser] = useState(null);
+
+  const signinWithGitHub = () => {
+    return firebase
+      .auth()
+      .signInWithPopup(new firebase.auth.GithubAuthProvider())
+      .then((response) => setUser(response.user));
+  };
+
+  const signout = () => {
+    return firebase
+      .auth()
+      .signOut()
+      .then(() => setUser(false));
+  };
+
+  useEffect(() => {
+    const unsubscribe = firebase.auth().onAuthStateChanged(handleUser);
+    return () => unsubscribe();
+  }, []);
+
+  return {
+    user,
+    signinWithGitHub,
+    signout
+  };
+}
+```
+
+If you don't want to create the interface yourself, you can use [FirebaseUI](https://github.com/firebase/firebaseui-web).
+
+<details open>
+  <summary><b>Examples</b></summary>
+  <ul>
+    <li><a href="https://github.com/vercel/next.js/tree/canary/examples/with-firebase-authentication">with-firebase-authentication</a></li>
+  </ul>
+</details
```

Comment:
We *always* put these at the top under the header

Reply:
I moved them underneath each section's header. Is this okay, or would you prefer them all grouped together?